wire-ios-request-strategy/Sources/Helpers/FederationDeleteManager.swift

```diff
@@ -0,0 +1,120 @@
+////
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+
+import Foundation
+import WireDataModel
+
+final public class FederationDeleteManager {
+    private weak var syncContext: NSManagedObjectContext?
+
+    public init(syncContext: NSManagedObjectContext? = nil) {
+        self.syncContext = syncContext
+    }
+
+    public func backendStoppedFederatingWithDomain(domain: String) {
+        guard let moc = syncContext,
+              let selfDomain = ZMUser.selfUser(in: moc).domain else { return }
+
+        markAllOneToOneConversationsAsReadOnly(forDomain: domain)
+        removeConnectionRequests(withDomain: domain)
+        domainsStoppedFederating(domains: [selfDomain, domain])
+    }
+
+    public func domainsStoppedFederating(domains: [String]) {
```

Comment:
Is it possible that the backend notifies clients than more than two domains stopped federating ? 

**if NO**: would it make sense to be more restrictive with the parameters and accept only two domains? I get that the array type is useful for certain operations such as `contains` or `filter`, but it could still be created from the two parameters once within the method

**if YES**: Then the logic when none of the conversations is owned by the domains to be removed might be flawed.

```swift
for currentConversation in conversations {
   if domains.contains(currentConversation.domain ?? "") {
      // ...
   } else { 
      // the logic from here
      guard currentConversation.containsParticipantsFromAllDomains(domains: domains) else { continue }
      removeParticipantsFromDomains(domains: domains, inConversation: currentConversation)
   }
}       
```

In case there are more than two backend that stop federating (let's say 3), then participants will only be removed from conversations that have participants from all 3 domains. But not from conversations that have participants from 2 unfederated domains, although these participants should not be able to federate anymore